data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala

```diff
@@ -191,65 +213,60 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
     val userLocationResolvedDF = userOrgDenormDF
       .join(locationDenormDF, Seq("userid"), "left_outer")
 
+    /**
+      * Resolve the state name:
+      * 1. State-Users:
+      * ORGANISATION.locationids=LOCATION.id and LOCATION.type='state'
+      * TO-DO: How to map it to the user table to get the respective user
+      *
+      * 2. Custodian User
+      * USER.locationids=LOCATION.id and LOCATION.type='state' and fetch the name,userid
+      */
+
+    val stateInfoByUser = userDF.withColumn("explode_location", explode(col("locationids")))
+      .join(locationDF, col("explode_location") === locationDF.col("id")
+        && locationDF.col("type") === "state")
+      .select(col("name").as("state_name"), col("userid"))
+
+    val locationInfoByOrg = organisationDF.withColumn("explode_location", explode(col("locationids")))
+      .join(locationDF, col("explode_location") === locationDF.col("id")
+        && locationDF.col("type") === "state")
+      .select(organisationDF.col("id"), locationDF.col("name").as("state_name"),
+        organisationDF.col("isrootorg"))
+
+    val stateInfoByOrg = locationInfoByOrg.join(userDF,
+      locationInfoByOrg.col("id") === userDF.col("rootorgid") &&
+        locationInfoByOrg.col("isrootorg").equalTo(true))
+      .select(col("state_name"),
+        userDF.col("userid"))
+
+    val resolvedStateDenormDF = stateInfoByUser.union(stateInfoByOrg)
+        .dropDuplicates(Seq("userid"))
+
```

Comment:
What is the purpose of this logic?

Reply:
Can we move this to same method - getUserSelfDeclaredDetails() & getStateDeclaredDetails()?

Reply:
Done